File: src/js/heap_manipulator.js
/**
 * JavaScript heap manipulation module for generating heap spray and defragmentation code.
 */

class SprayConfig {
    /**
     * Configuration for heap spray operations.
     * @param {number} targetSize - Target size for spray objects
     * @param {number} count - Number of objects to spray
     * @param {string} objectType - Type of object ('array', 'object', 'string')
     * @param {string} [fillPattern] - Optional fill pattern
     * @param {number} [alignment=8] - Memory alignment
     */
    constructor(targetSize, count, objectType, fillPattern = null, alignment = 8) {
        this.targetSize = targetSize;
        this.count = count;
        this.objectType = objectType;
        this.fillPattern = fillPattern;
        this.alignment = alignment;
    }
}

class HeapManipulator {
    constructor() {
        this.sprayHistory = [];
    }

    /**
     * Generate JavaScript code for heap spraying
     * @param {SprayConfig} config - Spray configuration
     * @returns {string} Generated JavaScript code
     */
    generateSprayCode(config) {
        this.sprayHistory.push(config);

        switch (config.objectType) {
            case 'array':
                return this._generateArraySpray(config);
            case 'object':
                return this._generateObjectSpray(config);
            case 'string':
                return this._generateStringSpray(config);
            default:
                throw new Error(`Unsupported object type: ${config.objectType}`);
        }
    }

    /**
     * Generate code for array-based heap spray
     * @private
     */
    _generateArraySpray(config) {
        const fillValue = config.fillPattern || '0x41';
        return `
const spray = new Array(${config.count});
for (let i = 0; i < spray.length; i++) {
    spray[i] = new Array(${config.targetSize}).fill(Number(${fillValue}));
}`;
    }

    /**
     * Generate code for object-based heap spray
     * @private
     */
    _generateObjectSpray(config) {
        const properties = this._generateObjectProperties(config.targetSize);
        return `
const spray = new Array(${config.count});
for (let i = 0; i < spray.length; i++) {
    spray[i] = {
        ${properties.map(prop => `${prop}: Number(${config.fillPattern || '0x41'})`).join(',\n        ')}
    };
}`;
    }

    /**
     * Generate code for string-based heap spray
     * @private
     */
    _generateStringSpray(config) {
        const fillChar = config.fillPattern || 'A';
        return `
const spray = new Array(${config.count});
for (let i = 0; i < spray.length; i++) {
    spray[i] = '${fillChar}'.repeat(${config.targetSize});
}`;
    }

    /**
     * Generate code for heap defragmentation
     * @param {number} targetSize - Target size for dummy objects
     * @param {number} [count=100] - Number of dummy objects
     * @returns {string} Generated JavaScript code
     */
    generateDefragCode(targetSize, count = 100) {
        return `
const dummy = [];
for (let i = 0; i < ${count}; i++) {
    dummy.push(new Array(${targetSize}).fill(Number(0x42)));
}

// Force garbage collection
dummy.length = 0;
if (global.gc) global.gc();`;
    }

    /**
     * Generate object properties to achieve target size
     * @private
     */
    _generateObjectProperties(targetSize) {
        const properties = [];
        let currentSize = 0;

        while (currentSize < targetSize) {
            const propName = this._generateRandomString(8);
            properties.push(propName);
            currentSize += 8; // Approximate property size
        }

        return properties;
    }

    /**
     * Generate random string of specified length
     * @private
     */
    _generateRandomString(length) {
        const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
        return Array.from({ length }, () => chars[Math.floor(Math.random() * chars.length)]).join('');
    }

    /**
     * Generate code to prime freelist for specific slot size
     * @param {number} slotSize - Size of slots to prime
     * @param {number} count - Number of objects to create
     * @returns {string} Generated JavaScript code
     */
    generateFreelistPriming(slotSize, count) {
        return `
const prime = new Array(${count});
for (let i = 0; i < prime.length; i++) {
    prime[i] = new Array(${slotSize}).fill(Number(0x43));
}

// Free them to prime freelist
prime.length = 0;
if (global.gc) global.gc();`;
    }
}

// Export for use in other modules
export { HeapManipulator, SprayConfig }; 

File: src/js/heap_manipulator.py
"""
JavaScript heap manipulation module for generating heap spray and defragmentation code.
"""

from typing import List, Dict, Optional
from dataclasses import dataclass
import random
import string

@dataclass
class SprayConfig:
    """Configuration for heap spray operations."""
    target_size: int
    count: int
    object_type: str  # 'array', 'object', 'string'
    fill_pattern: Optional[str] = None
    alignment: int = 8

class HeapManipulator:
    """Generates JavaScript code for heap manipulation."""
    
    def __init__(self):
        self.spray_history: List[SprayConfig] = []
        
    def generate_spray_code(self, config: SprayConfig) -> str:
        """Generate JavaScript code for heap spraying."""
        self.spray_history.append(config)
        
        if config.object_type == 'array':
            return self._generate_array_spray(config)
        elif config.object_type == 'object':
            return self._generate_object_spray(config)
        elif config.object_type == 'string':
            return self._generate_string_spray(config)
        else:
            raise ValueError(f"Unsupported object type: {config.object_type}")
            
    def _generate_array_spray(self, config: SprayConfig) -> str:
        """Generate code for array-based heap spray."""
        fill_value = config.fill_pattern or '0x41'
        code = []
        
        # Create array with specific size
        code.append(f"const spray = new Array({config.count});")
        code.append("for (let i = 0; i < spray.length; i++) {")
        code.append(f"    spray[i] = new Array({config.target_size}).fill(Number({fill_value}));")
        code.append("}")
        
        return "\n".join(code)
        
    def _generate_object_spray(self, config: SprayConfig) -> str:
        """Generate code for object-based heap spray."""
        code = []
        code.append(f"const spray = new Array({config.count});")
        
        # Generate object with specific shape
        properties = self._generate_object_properties(config.target_size)
        code.append("for (let i = 0; i < spray.length; i++) {")
        code.append("    spray[i] = {")
        for prop in properties:
            code.append(f"        {prop}: Number({config.fill_pattern or '0x41'}),")
        code.append("    };")
        code.append("}")
        
        return "\n".join(code)
        
    def _generate_string_spray(self, config: SprayConfig) -> str:
        """Generate code for string-based heap spray."""
        fill_char = config.fill_pattern or 'A'
        code = []
        
        code.append(f"const spray = new Array({config.count});")
        code.append("for (let i = 0; i < spray.length; i++) {")
        code.append(f"    spray[i] = '{fill_char}'.repeat({config.target_size});")
        code.append("}")
        
        return "\n".join(code)
        
    def generate_defrag_code(self, target_size: int, count: int = 100) -> str:
        """Generate code for heap defragmentation."""
        code = []
        
        # Create dummy objects to force GC
        code.append("const dummy = [];")
        code.append(f"for (let i = 0; i < {count}; i++) {{")
        code.append(f"    dummy.push(new Array({target_size}).fill(Number(0x42)));")
        code.append("}")
        
        # Force garbage collection
        code.append("dummy.length = 0;")
        code.append("if (global.gc) global.gc();")
        
        return "\n".join(code)
        
    def _generate_object_properties(self, target_size: int) -> List[str]:
        """Generate object properties to achieve target size."""
        properties = []
        current_size = 0
        
        while current_size < target_size:
            prop_name = ''.join(random.choices(string.ascii_letters, k=8))
            properties.append(prop_name)
            current_size += 8  # Approximate property size
            
        return properties
        
    def generate_freelist_priming(self, slot_size: int, count: int) -> str:
        """Generate code to prime freelist for specific slot size."""
        code = []
        
        # Create objects of specific size
        code.append(f"const prime = new Array({count});")
        code.append("for (let i = 0; i < prime.length; i++) {")
        code.append(f"    prime[i] = new Array({slot_size}).fill(Number(0x43));")
        code.append("}")
        
        # Free them to prime freelist
        code.append("prime.length = 0;")
        code.append("if (global.gc) global.gc();")
        
        return "\n".join(code) 